go/app/infra.go

```diff
@@ -17,180 +18,174 @@ var errItemNotFound = errors.New("item not found")
 	GetItemById(ctx context.Context, item_id string) (Item, error)
+	SearchItemsByKeyword(ctx context.Context, keyword string) ([]Item, error)
 }
 
-// itemRepository is an implementation of ItemRepository
 type itemRepository struct {
-	// fileName is the path to the JSON file storing items.
-	fileName string
+	db *sql.DB
 }
 
-// NewItemRepository creates a new itemRepository.
-// main.goを実行するディレクトリによってfileNameを変更する
-func NewItemRepository() ItemRepository {
-	return &itemRepository{fileName: "cmd/api/items.json"}
+// 返り値を増やした
+// -> server.goのRun()でNewItemRepositoryのerrを検知できずに
+// nilのitemRepoを使用したことによるnil参照panicを防ぐ
+func NewItemRepository(db *sql.DB) (ItemRepository, error) {
+	// items tableがなかったら作成
+	q, err := os.ReadFile("db/items.sql")
+	if err != nil {
+		return &itemRepository{}, err
+	}
+
+	query := string(q)
+	_, err = db.Exec(query)
+	if err != nil {
+		slog.Error("failed to create items table and categories table", "error", err)
+		return nil, err
+	}
+
+	return &itemRepository{db: db}, nil
 }
 
-// Insert inserts an item into the repository.
 func (i *itemRepository) Insert(ctx context.Context, item *Item) error {
-	// STEP 4-2: add an implementation to store an item
-	// /api内で go run main.go(パスを"items.json"でやるなら)
+	var categoryID int
 
-	// jsonファイルを読み込み
-	// ここでエラーだったらjsonFileには空のバイト列が格納される
-	jsonFile, err := os.ReadFile(i.fileName)
-	// error処理
+	// categories tableから(categories tableの)name = item.Categoryのidを取得
+	err := i.db.QueryRow("SELECT id FROM categories WHERE name = ?", item.Category).Scan(&categoryID)
 	if err != nil {
-		if errors.Is(err, os.ErrNotExist) {
-			// ファイルが存在しなかったら作成
-			file, err := os.Create(i.fileName)
+		if err == sql.ErrNoRows {
+			// 該当する行がなかったら = 新しいカテゴリーだったら
+			// categories tableのnameにitem.Categoryの値をinsert
+			res, err := i.db.Exec("INSERT INTO categories (name) VALUES (?)", item.Category)
 			if err != nil {
 				return err
 			}
-			defer file.Close()
-			defaultJson := []byte(`{"items":[]}`)
-			_, err = file.Write(defaultJson)
+			// 最後に挿入された自動採番(AUTOINCREMENT)のidを取得
+			id, err := res.LastInsertId()
 			if err != nil {
 				return err
 			}
+			categoryID = int(id)
 		} else {
-			// その他のエラーだったら中断、エラーを返す
-			println("Error:", err)
 			return err
 		}
 	}
-	// os.ErrNotExistでデフォルトデータが書き込まれたらスキップされる
-	if len(jsonFile) == 0 {
-		jsonFile = []byte(`{"items":[]}`)
-	}
-
-	// jsonから構造体に変換
-	// jsonのitems配列の各要素がItem構造体として格納される
-	var data struct {
-		Items []Item `json:"items"`
-	}
-	if err = json.Unmarshal(jsonFile, &data); err != nil {
-		return err
-	}
-
-	// itemsにitemを追加
-	data.Items = append(data.Items, *item)
-
-	// 構造体からjsonに変換
-	itemsJson, err := json.MarshalIndent(data, "", "  ")
-	if err != nil {
-		return err
-	}
-
-	// jsonファイルに書き込み
-	// 0644 ファイルのパーミッション
-	// os.ErrNotExistで作成されたファイルと既に存在していたファイルは
-	// どちらにせよここで同じパスに書き込まれる(i.fileNameで指定しているから)
-	err = os.WriteFile(i.fileName, itemsJson, 0644)
+	// insert
+	query := "INSERT INTO items (name, category_id, image_name) VALUES (?, ?, ?)"
```

Comment:
Instead of querying the `category` separately, we can use join like everywhere else in the original query (line 65) above